### PR TITLE
Encrypt pbft ledger

### PR DESCRIPTION
**Pull request by olgavrou**
IV for the PBFT case is now constructed by the `view/term` and `kv version` instead of `node id` and a monotonically increasing sequence number.

For the view to be updated in PBFT need to set it via the encryptor when there is a view change.

### CODE REVIEW

**Comment by cimetrics** [comment]
ledger_encrypt_pbft@[5870 aka 20200310.22](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5870&view=results) vs master ewma over 30 builds from [5516](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5516&view=results) to [5855](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5855&view=results)
images

File: src/consensus/raft/raft.h [hunk 1]
```diff
@@ -132,6 +134,8 @@ namespace raft
       NodeId id,
       std::chrono::milliseconds request_timeout_,
       std::chrono::milliseconds election_timeout_,
+      std::shared_ptr<kv::AbstractTxEncryptor> encryptor_ =
```
Comment (jumaffre) [inline src/consensus/raft/raft.h]: I wouldn't expect the `NullTxEncryptor` to be used here. Could this be `nullptr` instead?

Comment (olgavrou) [reply src/consensus/raft/raft.h]: For testing purposes it is nicer to set a NullTxEncryptor by default instead of going to all of our unit tests and setting one manually. It also saves a lot of nullptr checks

File: src/consensus/pbft/pbft.h [hunk 2]
```diff
@@ -341,11 +347,13 @@ namespace pbft
         if (*gb_info->last_commit_view < view)
         {
           gb_info->view_change_list->emplace_back(view, version);
+          gb_info->encryptor->set_view(view);
```
Comment (olgavrou) [inline src/consensus/pbft/pbft.h]: This is not correct, we need to set the view on the encryptor right after we view change. Will change this.

File: src/node/encryptor.h [hunk 3]
```diff
@@ -85,10 +85,11 @@ namespace ccf
 
     std::list<EncryptionKey> encryption_keys;
 
-    void set_iv(crypto::GcmHeader<crypto::GCM_SIZE_IV>& gcm_hdr)
+    void set_iv(
+      crypto::GcmHeader<crypto::GCM_SIZE_IV>& gcm_hdr, kv::Version version)
     {
-      gcm_hdr.set_iv_id(id);
-      gcm_hdr.set_iv_seq(seqNo.fetch_add(1));
+      gcm_hdr.set_iv_view(view);
```
Comment (ashamis) [inline src/node/encryptor.h]: Is this safe? This looks like a problem with multi-threading

File: src/node/encryptor.h [hunk 4]
```diff
@@ -57,10 +58,9 @@ namespace ccf
   class TxEncryptor : public kv::AbstractTxEncryptor
   {
   private:
-    NodeId id;
     bool is_recovery;
 
-    std::atomic<SeqNo> seqNo{0};
+    View view;
```
Comment (ashamis) [inline src/node/encryptor.h]: Why does this not need to be atomic?

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 5]
```diff
@@ -391,12 +391,8 @@ bool Replica::compare_execution_results(
 
   if (!execution_match)
   {
-    if (rollback_cb != nullptr)
-    {
-      rollback_cb(last_te_version, rollback_info);
-    }
-    last_tentative_execute--;
-    return false;
+    throw std::logic_error(
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/Replica.cpp]: assert only on playback

File: src/consensus/raft/raft.h (hunk 1 above)
Comment (jumaffre) [reply src/consensus/raft/raft.h]: I think that using the `NullTxEncryptor` as default should be avoided, and checking for `nullptr` is fine (it's only required once in this file).

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp [hunk 6]
```diff
@@ -376,7 +376,9 @@ TEST_CASE("Test Ledger Replay")
         REQUIRE(
           store->deserialise_views(corrupt_entry, false, nullptr, &tx) ==
           kv::DeserialiseSuccess::PASS_PRE_PREPARE);
-        pbft::GlobalState::get_replica().playback_pre_prepare(tx);
+        REQUIRE_THROWS_AS(
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: Why is this required?

File: src/consensus/pbft/pbft.h [hunk 7]
```diff
@@ -241,15 +248,18 @@ namespace pbft
       pbft::PrePreparesMap& pbft_pre_prepares_map,
       const std::string& privk_pem,
       const std::vector<uint8_t>& cert,
-      const consensus::Config& consensus_config) :
+      const consensus::Config& consensus_config,
+      std::shared_ptr<kv::AbstractTxEncryptor> encryptor_ =
```
Comment (jumaffre) [inline src/consensus/pbft/pbft.h]: Same comment as for Raft here.

File: src/crypto/symmkey.h [hunk 8]
```diff
@@ -34,10 +34,10 @@ namespace crypto
       memcpy(iv, data.data() + sizeof(tag), sizeof(iv));
     }
 
-    void set_iv_id(uint64_t id)
+    void set_iv_view(uint64_t view)
```
Comment (jumaffre) [inline src/crypto/symmkey.h]: Just to check: in PBFT, does the view number start from 0 or 1?

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp (hunk 6 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: Because we now throw if when we are playing back the merkle roots do not match

File: src/node/encryptor.h [hunk 9]
```diff
@@ -85,10 +88,33 @@ namespace ccf
...
+    void check_iv_validity(kv::Version version, View view)
     {
-      gcm_hdr.set_iv_id(id);
-      gcm_hdr.set_iv_seq(seqNo.fetch_add(1));
+      // expects iv to be locked
+      if (version <= last_used_version)
+      {
+        if (view <= last_used_view)
+        {
+          throw std::logic_error(fmt::format(
```
Comment (jumaffre) [inline src/node/encryptor.h]: Should we safely recovery from this?

Comment (olgavrou) [reply src/node/encryptor.h]: How would you propose we recover?

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp (hunk 6 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: In normal execution, when merkle roots do not match we expect to view change. When we are playing back we can't view change so the only solution is for the node to start again.

File: src/crypto/symmkey.h (hunk 8 above)
Comment (olgavrou) [reply src/crypto/symmkey.h]: in PBFT view starts from 0

File: src/node/encryptor.h (hunk 9 above)
Comment (olgavrou) [reply src/node/encryptor.h]: Will remove this check. Will keep the original implementation for Raft and a view + version implementation for PBFT. It should be impossible for PBFT to rollback without a view change, so versions should not be re-used with the same view number. Hence this check is redundant.

File: src/kv/test/kv_bench.cpp [hunk 10]
```diff
@@ -62,7 +62,7 @@ static void deserialise(picobench::state& s)
   Store kv_store2;
 
   auto secrets = create_ledger_secrets();
-  auto encryptor = std::make_shared<ccf::TxEncryptor>(0x1, secrets);
+  auto encryptor = std::make_shared<ccf::RaftTxEncryptor>(0x1, secrets);
```
Comment (achamayou) [inline src/kv/test/kv_bench.cpp]: Why 0x1 here and 1 in the next file?

Comment (olgavrou) [reply src/kv/test/kv_bench.cpp]: I don't know it was like this when I found it. If no one can remember why I will change it to 1 here too.

File: src/node/nodestate.h [hunk 11]
```diff
@@ -763,14 +763,13 @@ namespace ccf
 
       if (network.consensus_type == ConsensusType::PBFT)
       {
-        // for now do not encrypt the ledger as the current implementation does
-        // not work for PBFT
-        recovery_encryptor = std::make_shared<NullTxEncryptor>();
+        recovery_encryptor =
+          std::make_shared<PbftTxEncryptor>(network.ledger_secrets, true);
       }
       else
```
Comment (ashamis) [inline src/node/nodestate.h]: can you please put else if RAFT here and then have an else case where we have some kind of error

Comment (olgavrou) [reply src/node/nodestate.h]: Done

File: src/consensus/pbft/pbfttypes.h [hunk 12]
```diff
@@ -148,6 +149,19 @@ namespace pbft
...
+
+    std::shared_ptr<kv::AbstractTxEncryptor> get_encryptor()
+    {
+      while (true)
+      {
+        auto p = x.lock();
+        if (p)
+        {
+          return p->get_encryptor();
+          break;
```
Comment (ashamis) [inline src/consensus/pbft/pbfttypes.h]: you do not need the break here

File: src/kv/test/kv_bench.cpp (hunk 10 above)
Comment (achamayou) [reply src/kv/test/kv_bench.cpp]: It's the same value, it's just... odd? :)
